spec/unit/util_spec.rb

```diff
@@ -24,48 +24,115 @@ def get_mode(file)
-      Puppet::Util.withenv @new_env do
-        expect(ENV["PATH"]).to eq("/some/bogus/path")
+    describe "default mode" do
+      before :each do
+        @original_path = ENV["PATH"]
+        @new_env = {:PATH => "/some/bogus/path"}
       end
-      expect(ENV["PATH"]).to eq(@original_path)
-    end
 
-    it "should reset environment variables to their original values even if the block fails" do
-      begin
+      it "should change environment variables within the block then reset environment variables to their original values" do
         Puppet::Util.withenv @new_env do
           expect(ENV["PATH"]).to eq("/some/bogus/path")
-          raise "This is a failure"
         end
-      rescue
+        expect(ENV["PATH"]).to eq(@original_path)
       end
-      expect(ENV["PATH"]).to eq(@original_path)
-    end
 
-    it "should reset environment variables even when they are set twice" do
-      # Setting Path & Environment parameters in Exec type can cause weirdness
-      @new_env["PATH"] = "/someother/bogus/path"
-      Puppet::Util.withenv @new_env do
-        # When assigning duplicate keys, can't guarantee order of evaluation
-        expect(ENV["PATH"]).to match(/\/some.*\/bogus\/path/)
+      it "should reset environment variables to their original values even if the block fails" do
+        begin
+          Puppet::Util.withenv @new_env do
+            expect(ENV["PATH"]).to eq("/some/bogus/path")
+            raise "This is a failure"
+          end
+        rescue
+        end
+        expect(ENV["PATH"]).to eq(@original_path)
       end
-      expect(ENV["PATH"]).to eq(@original_path)
-    end
 
-    it "should remove any new environment variables after the block ends" do
-      @new_env[:FOO] = "bar"
-      ENV["FOO"] = nil
-      Puppet::Util.withenv @new_env do
-        expect(ENV["FOO"]).to eq("bar")
+      it "should reset environment variables even when they are set twice" do
+        # Setting Path & Environment parameters in Exec type can cause weirdness
+        @new_env["PATH"] = "/someother/bogus/path"
+        Puppet::Util.withenv @new_env do
+          # When assigning duplicate keys, can't guarantee order of evaluation
+          expect(ENV["PATH"]).to match(/\/some.*\/bogus\/path/)
+        end
+        expect(ENV["PATH"]).to eq(@original_path)
+      end
+
+      it "should remove any new environment variables after the block ends" do
+        @new_env[:FOO] = "bar"
+        ENV["FOO"] = nil
+        Puppet::Util.withenv @new_env do
+          expect(ENV["FOO"]).to eq("bar")
+        end
+        expect(ENV["FOO"]).to eq(nil)
       end
-      expect(ENV["FOO"]).to eq(nil)
     end
 
+    describe "windows mode", :if => Puppet.features.microsoft_windows? do
+      before :each do
+        @original_path = ENV["PATH"]
+        @new_env = {:PATH => "/some/bogus/path"}
+      end
+
+      it "should preserve UTF8 environment variables" do
+        env_var_name = SecureRandom.uuid.force_encoding(Encoding::UTF_8)
```

Comment:
I don't think you need `force_encoding` here:

```
[2] pry(main)> f = SecureRandom.uuid
=> "1aef9610-d9fb-4c06-9627-72ef578efe95"
[3] pry(main)> f.encoding
=> #<Encoding:UTF-8>
```

`force_encoding` is generally something that you *don't* want to call explicitly - it tells Ruby to ignore the current encoding of the string that you have and to pretend that the bytes are representative of another encoding.  In this case, Ruby has generated a string, which by default is `UTF-8`.  (If you really wanted to *change* the encoding of the string to something else, you would call `.encode(Encoding::UTF_8)`)

We do need `force_encoding` a couple of lines later because we have bytes that we know are UTF-8 and that we ask Ruby to pack as a string... but by default, it doesn't know that:

```ruby
[5] pry(main)> s = [0xE2, 0x80, 0x93, 0xE2, 0x84, 0xA2].pack('c*')
=> "\xE2\x80\x93\xE2\x84\xA2"
[6] pry(main)> s.encoding
=> #<Encoding:ASCII-8BIT>

[7] pry(main)> s.force_encoding('utf-8')
=> "\u2013\u2122"
```